File: backend/apps/synie_core/lib/synie_core/authz/actor.ex
defmodule SynieCore.Authz.Actor do
  @moduledoc """
  请求期权限主体,作为 Ash actor 使用。

  在登录解析后由 `SynieCore.Authz.build_actor/1` 构建,
  携带用户的权限码集合(含通配)与授权公司范围。
  """

  @enforce_keys [:user_id]
  defstruct [
    :user_id,
    :username,
    super_admin: false,
    all_companies: false,
    permissions: MapSet.new(),
    company_ids: []
  ]

  @type t :: %__MODULE__{
          user_id: String.t(),
          username: String.t() | nil,
          super_admin: boolean(),
          all_companies: boolean(),
          permissions: MapSet.t(String.t()),
          company_ids: [String.t()]
        }
end

File: server/internal/platform/authz/actor.go
package authz

import (
	"slices"

	"github.com/google/uuid"
)

type Actor struct {
	UserID       uuid.UUID
	Username     string
	Name         *string
	SuperAdmin   bool
	AllCompanies bool
	Permissions  map[string]struct{}
	CompanyIDs   []uuid.UUID
}

func (a *Actor) HasPermission(code string) bool {
	if a == nil {
		return false
	}
	if a.SuperAdmin {
		return true
	}
	return Matches(a.Permissions, code)
}

func (a *Actor) CompanyFilter() (bypass bool, ids []uuid.UUID) {
	if a == nil {
		return false, nil
	}
	if a.SuperAdmin || a.AllCompanies {
		return true, nil
	}
	return false, slices.Clone(a.CompanyIDs)
}

func (a *Actor) CanAccessCompany(id uuid.UUID) bool {
	if a == nil {
		return false
	}
	if a.SuperAdmin || a.AllCompanies {
		return true
	}
	return slices.Contains(a.CompanyIDs, id)
}
